File: tests/RopeManager.test.mjs
import { describe, it, beforeEach } from "node:test";
import assert from "node:assert/strict";
import { __reset } from "./stubs/minecraft-server.mjs";
import { RopeManager } from "../ropes_bp/scripts/RopeManager.js";

describe("RopeManager", () => {
  let mgr;

  beforeEach(() => {
    __reset();
    mgr = new RopeManager();
  });

  it("createChain stores a chain retrievable by getChainAtPosition", () => {
    const chain = mgr.createChain("rope", "minecraft:overworld", { x: 0, y: 64, z: 0 }, "north", 10);
    assert.equal(chain.id, "rope_1");
    assert.equal(chain.type, "rope");
    assert.equal(chain.totalSegments, 10);

    const found = mgr.getChainAtPosition("minecraft:overworld", { x: 0, y: 64, z: 0 });
    assert.equal(found, chain);
  });

  it("save/load round-trips correctly", () => {
    mgr.createChain("rope", "minecraft:overworld", { x: 0, y: 64, z: 0 }, "north", 10);
    mgr.createChain("rope_ladder", "minecraft:overworld", { x: 5, y: 64, z: 5 }, "east", 5);

    const mgr2 = new RopeManager();
    mgr2.load();

    const chain1 = mgr2.getChain("rope_1");
    assert.ok(chain1);
    assert.equal(chain1.type, "rope");
    assert.equal(chain1.totalSegments, 10);

    const chain2 = mgr2.getChain("rope_2");
    assert.ok(chain2);
    assert.equal(chain2.type, "rope_ladder");
  });

  it("load is idempotent — calling twice does not duplicate chains", () => {
    mgr.createChain("rope", "minecraft:overworld", { x: 0, y: 64, z: 0 }, "north", 5);

    const mgr2 = new RopeManager();
    mgr2.load();
    mgr2.load();

    let count = 0;
    for (const _ of mgr2.getAllChains()) count++;
    assert.equal(count, 1);
  });

  it("indexes support block position in _supportIndex", () => {
    mgr.createChain("rope", "minecraft:overworld", { x: 5, y: 64, z: 5 }, "north", 5);
    const supportChains = mgr.getChainsForSupportBlock("minecraft:overworld", { x: 5, y: 64, z: 4 });
    assert.ok(supportChains);
    assert.ok(supportChains.has("rope_1"));
  });

  it("two chains on same support block (different faces) -> Set contains both", () => {
    // Both anchors on different faces of the block at {5, 64, 5}
    mgr.createChain("rope", "minecraft:overworld", { x: 5, y: 64, z: 4 }, "south", 5); // support at z+1 = {5,64,5}
    mgr.createChain("rope", "minecraft:overworld", { x: 4, y: 64, z: 5 }, "east", 5);  // support at x+1 = {5,64,5}
    const supportChains = mgr.getChainsForSupportBlock("minecraft:overworld", { x: 5, y: 64, z: 5 });
    assert.ok(supportChains);
    assert.ok(supportChains.has("rope_1"));
    assert.ok(supportChains.has("rope_2"));
  });

  it("removeChain cleans up position index and support index", () => {
    const chain = mgr.createChain("rope", "minecraft:overworld", { x: 0, y: 64, z: 0 }, "north", 5);
    mgr.removeChain(chain.id);

    assert.equal(mgr.getChainAtPosition("minecraft:overworld", { x: 0, y: 64, z: 0 }), null);
    assert.equal(mgr.getChainsForSupportBlock("minecraft:overworld", { x: 0, y: 64, z: -1 }), null);
  });

  it("position index uses dimension-prefixed keys — no cross-dimension collision", () => {
    mgr.createChain("rope", "minecraft:overworld", { x: 0, y: 64, z: 0 }, "up", 5);
    mgr.createChain("rope", "minecraft:nether", { x: 0, y: 64, z: 0 }, "up", 3);

    const ow = mgr.getChainAtPosition("minecraft:overworld", { x: 0, y: 64, z: 0 });
    const nether = mgr.getChainAtPosition("minecraft:nether", { x: 0, y: 64, z: 0 });

    assert.equal(ow.id, "rope_1");
    assert.equal(nether.id, "rope_2");
  });

  it("addSegmentPosition / removeSegmentPosition updates index", () => {
    const chain = mgr.createChain("rope", "minecraft:overworld", { x: 0, y: 64, z: 0 }, "up", 5);
    mgr.addSegmentPosition(chain.id, "minecraft:overworld", { x: 0, y: 63, z: 0 });

    assert.equal(mgr.getChainAtPosition("minecraft:overworld", { x: 0, y: 63, z: 0 }), chain);

    mgr.removeSegmentPosition("minecraft:overworld", { x: 0, y: 63, z: 0 });
    assert.equal(mgr.getChainAtPosition("minecraft:overworld", { x: 0, y: 63, z: 0 }), null);
  });

  it("getChain returns null for unknown id", () => {
    assert.equal(mgr.getChain("nonexistent"), null);
  });

  it("getChainAtPosition returns null for unindexed position", () => {
    assert.equal(mgr.getChainAtPosition("minecraft:overworld", { x: 99, y: 99, z: 99 }), null);
  });
});

File: java-ropes/src/test/java/com/ropes/domain/RopeManagerTest.java
package com.ropes.domain;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

/** Port of tests/RopeManager.test.mjs. Persistence is exercised via the
 *  exportState/load map exchange that stands in for the world property I/O. */
class RopeManagerTest {

    private RopeManager mgr;

    private static RopePos p(int x, int y, int z) {
        return new RopePos(x, y, z);
    }

    @BeforeEach
    void setUp() {
        mgr = new RopeManager();
    }

    @Test
    void createChainStoresChainRetrievableByPosition() {
        RopeChain chain = mgr.createChain("rope", "minecraft:overworld", p(0, 64, 0), "north", 10);
        assertEquals("rope_1", chain.getId());
        assertEquals("rope", chain.getType());
        assertEquals(10, chain.getTotalSegments());

        RopeChain found = mgr.getChainAtPosition("minecraft:overworld", p(0, 64, 0));
        assertSame(chain, found);
    }

    @Test
    void saveLoadRoundTrips() {
        mgr.createChain("rope", "minecraft:overworld", p(0, 64, 0), "north", 10);
        mgr.createChain("rope_ladder", "minecraft:overworld", p(5, 64, 5), "east", 5);
        List<Map<String, Object>> exported = mgr.exportState();

        RopeManager mgr2 = new RopeManager();
        mgr2.load(exported);

        RopeChain chain1 = mgr2.getChain("rope_1");
        assertNotNull(chain1);
        assertEquals("rope", chain1.getType());
        assertEquals(10, chain1.getTotalSegments());

        RopeChain chain2 = mgr2.getChain("rope_2");
        assertNotNull(chain2);
        assertEquals("rope_ladder", chain2.getType());
    }

    @Test
    void loadIsIdempotent() {
        mgr.createChain("rope", "minecraft:overworld", p(0, 64, 0), "north", 5);
        List<Map<String, Object>> exported = mgr.exportState();

        RopeManager mgr2 = new RopeManager();
        mgr2.load(exported);
        mgr2.load(exported);

        assertEquals(1, mgr2.getAllChains().size());
    }

    @Test
    void indexesSupportBlockPosition() {
        mgr.createChain("rope", "minecraft:overworld", p(5, 64, 5), "north", 5);
        Set<String> supportChains = mgr.getChainsForSupportBlock("minecraft:overworld", p(5, 64, 4));
        assertNotNull(supportChains);
        assertTrue(supportChains.contains("rope_1"));
    }

    @Test
    void twoChainsOnSameSupportBlockDifferentFaces() {
        mgr.createChain("rope", "minecraft:overworld", p(5, 64, 4), "south", 5); // support z+1 = (5,64,5)
        mgr.createChain("rope", "minecraft:overworld", p(4, 64, 5), "east", 5);  // support x+1 = (5,64,5)
        Set<String> supportChains = mgr.getChainsForSupportBlock("minecraft:overworld", p(5, 64, 5));
        assertNotNull(supportChains);
        assertTrue(supportChains.contains("rope_1"));
        assertTrue(supportChains.contains("rope_2"));
    }

    @Test
    void removeChainCleansUpIndices() {
        RopeChain chain = mgr.createChain("rope", "minecraft:overworld", p(0, 64, 0), "north", 5);
        mgr.removeChain(chain.getId());

        assertNull(mgr.getChainAtPosition("minecraft:overworld", p(0, 64, 0)));
        assertNull(mgr.getChainsForSupportBlock("minecraft:overworld", p(0, 64, -1)));
    }

    @Test
    void positionIndexUsesDimensionPrefixedKeys() {
        mgr.createChain("rope", "minecraft:overworld", p(0, 64, 0), "up", 5);
        mgr.createChain("rope", "minecraft:nether", p(0, 64, 0), "up", 3);

        RopeChain ow = mgr.getChainAtPosition("minecraft:overworld", p(0, 64, 0));
        RopeChain nether = mgr.getChainAtPosition("minecraft:nether", p(0, 64, 0));

        assertEquals("rope_1", ow.getId());
        assertEquals("rope_2", nether.getId());
    }

    @Test
    void addAndRemoveSegmentPositionUpdatesIndex() {
        RopeChain chain = mgr.createChain("rope", "minecraft:overworld", p(0, 64, 0), "up", 5);
        mgr.addSegmentPosition(chain.getId(), "minecraft:overworld", p(0, 63, 0));
        assertSame(chain, mgr.getChainAtPosition("minecraft:overworld", p(0, 63, 0)));

        mgr.removeSegmentPosition("minecraft:overworld", p(0, 63, 0));
        assertNull(mgr.getChainAtPosition("minecraft:overworld", p(0, 63, 0)));
    }

    @Test
    void getChainReturnsNullForUnknownId() {
        assertNull(mgr.getChain("nonexistent"));
    }

    @Test
    void getChainAtPositionReturnsNullForUnindexedPosition() {
        assertNull(mgr.getChainAtPosition("minecraft:overworld", p(99, 99, 99)));
    }
}
